### PR TITLE
Kaspervan m view results

### LINKED ISSUE #57: View results
### Description

Now, results are not visible to anyone.

### Solution

When a task is done, the people should be redirected to a thank you page. On the page is a button which will activate the download of the results in json format. When the results are private, first the user is prompted to enter a password.

New elements:
1. Thank you page with link to results;
2. Passwords in tasks;
3. Results file generated.

### PR DESCRIPTION
**Pull request by KaspervanM**
### Description

This pull request will close #57.
In this PR, a new page is added to which users will be redirected when a task has been completed. 

For this feature, a change to the taskloop has been made so the cookies are no longer removed when a task has been completed, this enables the redirect to the thank you-/ result page.

The result page has also been added to the router index file.

There has been a change in the component of the progress bar so it emits a signal when it hits 100%. This signal is picked up by the dashboard and redirects the user to the thank you-page. Also, its initial value is now 0% instead of 60%.

There has been a change to the server and taskservice: A get request for results has been added. This is used by the result page to get the results and to put them in a file that then gets downloaded. The download is initiated after the user presses a button, which will provide a confirm prompt (as a placeholder for a future password prompt).  When the user selects *yes*, the download starts.

Note that not all of issue #57 is solved in this PR yet. 
Still needed:
- [x] Password protected tasks
- [x] A way for the cookies to be removed after the task has finished

### References

* For the file download:
  https://stackoverflow.com/questions/53772331/vue-html-js-how-to-download-a-file-to-browser-using-the-download-tag
* For the button:
  https://bootstrap-vue.org/docs/components/button

### CODE REVIEW

**Comment by KaspervanM** [comment]
I have added the functionality, but the error handling is not there yet. If a wrong password gets sent to the server, it sends 0 as result. This should obviously not be the case. But in this branch the old error handling system is still used. This should eventually be changed, but for now, everything is operational.

File: server/src/app.ts [hunk 1]
```diff
@@ -232,17 +238,32 @@ app.post("/task", (req, res) => {
   return res.status(400).send(errmsg).end();
 }); //Add one task (JSON: request contains title, description and code, response contains id, title, description and code)
 
+app.post("/task/results", (req, res) => {
+  if (tasks[req.body.id].config["PRIVATE"]) {
+    if (tasks[req.body.id].password != md5(req.body.password))
```
Comment (evanpacini) [inline server/src/app.ts:243]:
```suggestion
    if(!req.body.password) return res.sendStatus(401); // Unauthorized
    if (tasks[req.body.id].password != md5(req.body.password)) return res.sendStatus(403); //Forbidden
```

File: server/src/app.ts [hunk 2]
```diff
@@ -232,17 +238,32 @@ app.post("/task", (req, res) => {
...
+      return res.send([0]);
```
Comment (evanpacini) [inline server/src/app.ts:244]:
```suggestion
```

File: server/src/app.ts [hunk 3]
```diff
@@ -232,17 +238,32 @@ app.post("/task", (req, res) => {
...
+  }
```
Comment (evanpacini) [inline server/src/app.ts:245]:
```suggestion
```

File: server/src/app.ts [hunk 4]
```diff
@@ -4,6 +4,7 @@ import { v4 as uuidv4 } from "uuid";
 const app = express();
 const port = 3000;
 const cors = require("cors");
+const md5 = require("md5");
```
Comment (evanpacini) [inline server/src/app.ts:7]: Don't use md5. It's extremely unsafe and should NEVER be used. It was cracked and a lot of rainbow tables are online for md5. SHA-1 used to be a good replacement but it's officially considered vulnerable too. I suggest using SHA-256, SHA-384 or SHA-512. These are safe and will work easily with JS.

File: server/src/app.ts [hunk 5]
```diff
@@ -232,17 +238,32 @@ app.post("/task", (req, res) => {
...
+    req.body.task.password &&
     req.body.task.title &&
     req.body.task.description &&
     req.body.task.config &&
     req.body.task.code &&
     req.body.task.id in tasks
   ) {
+    if (tasks[req.body.id].config["PRIVATE"]) {
+      if (tasks[req.body.id].password != md5(req.body.password))
+        return res.send(<Task>{});
```
Comment (evanpacini) [inline server/src/app.ts:261]:
```suggestion
        return res.sendStatus(403); // Forbidden
```

File: server/src/app.ts [hunk 6]
```diff
@@ -266,6 +287,10 @@ app.put("/task", (req, res) => {
 }); //Update one task (JSON: request contains id, title, description and code, response contains id, title, description and code)
 
 app.delete("/task/:id", (req, res) => {
+  if (tasks[req.body.id].config["PRIVATE"]) {
+    if (tasks[req.body.id].password != md5(req.body.password))
+      return res.send(<Task>{});
```
Comment (evanpacini) [inline server/src/app.ts:292]:
```suggestion
      return res.sendStatus(403); // Forbidden
```

File: server/src/app.ts [hunk 7]
```diff
@@ -232,17 +238,32 @@ app.post("/task", (req, res) => {
...
+    if (tasks[req.body.id].password != md5(req.body.password))
+      return res.send([0]);
+  }
+  const id = req.body.id;
+  return res.send([tasks[id].result]);
+});
+
 app.put("/task", (req, res) => {
   if (
+    req.body.task.password &&
```
Comment (evanpacini) [inline server/src/app.ts:252]: Only required if the task was private.
```suggestion
```

File: server/src/app.ts [hunk 8]
```diff
@@ -232,17 +238,32 @@ app.post("/task", (req, res) => {
...
     req.body.task.title &&
     req.body.task.description &&
     req.body.task.config &&
     req.body.task.code &&
     req.body.task.id in tasks
   ) {
+    if (tasks[req.body.id].config["PRIVATE"]) {
+      if (tasks[req.body.id].password != md5(req.body.password))
```
Comment (evanpacini) [inline server/src/app.ts:260]:
```suggestion
      if (!req.body.password) return res.sendStatus(401); // Unauthorized
      if (tasks[req.body.id].password != md5(req.body.password))
```

File: server/src/app.ts [hunk 9]
```diff
@@ -266,6 +287,10 @@ app.put("/task", (req, res) => {
 }); //Update one task (JSON: request contains id, title, description and code, response contains id, title, description and code)
 
 app.delete("/task/:id", (req, res) => {
+  if (tasks[req.body.id].config["PRIVATE"]) {
+    if (tasks[req.body.id].password != md5(req.body.password))
```
Comment (evanpacini) [inline server/src/app.ts:291]:
```suggestion
    if (!req.body.password) return res.sendStatus(401); // Unauthorized
    if (tasks[req.body.id].password != md5(req.body.password))
```

**Review by evanpacini — CHANGES_REQUESTED** [review]
The overall idea looks good, but the implementation is a bit lacking. MD5 should definitely NOT be used, in favor of SHA-256, 384 or 512. The overall code is outdated, which needs to be updated and following the new standard to be merged. Also there is little to no error handling, which needs to be fixed. I made some recommendations in the server code, but, that's following the pattern I made in my optimisation process. So that will only apply to when this branch is up to date. Also there are many conflicts in this branch, due to it being based on out-of-date code. So I recommend making a new branch based on master, and copying the core functionalities, improving upon aforementioned points.

**Comment by KaspervanM** [comment]
Yes I agree. I was considering making a new branch as well. However, before i do so, I'd like to know what you think of the functionality. I know the code needs to be changed, but does anything else need to be changed or added?

Also, I'd like to know if the downloading system works on your device, I had noticed some irregularities, but if all is OK on your device, I think it might be a browser setting.

**Comment by evanpacini** [comment]
I think the functionality is great, even though it's still unclear to me what "PRIVATE" means. As of now, it only means there is a password when you edit/download the results, but you can still find the task in the ViewTask, and anyone can still help the task. Maybe that should be clarified or changed. I tested the download result and that works just fine on my phone and on your laptop. That shouldn't be a problem.

**Comment by KaspervanM** [comment]
> I think the functionality is great, even though it's still unclear to me what "PRIVATE" means. As of now, it only means there is a password when you edit/download the results, but you can still find the task in the ViewTask, and anyone can still help the task. Maybe that should be clarified or changed. I tested the download result and that works just fine on my phone and on your laptop. That shouldn't be a problem.

Private means the result is only available if you have a password. Not allowing anonymous users to help or view a task is set in allow_anonymous_users. Although, this is not yet applied.

Perhaps the keyword should change, because 'private' is indeed unclear. What do you think would be clearer? Something along the lines of PRIVATE_RESULT perhaps or the other way around, PUBLIC_RESULT.

**Comment by evanpacini** [comment]
> > I think the functionality is great, even though it's still unclear to me what "PRIVATE" means. As of now, it only means there is a password when you edit/download the results, but you can still find the task in the ViewTask, and anyone can still help the task. Maybe that should be clarified or changed. I tested the download result and that works just fine on my phone and on your laptop. That shouldn't be a problem.
> 
> Private means the result is only available if you have a password. Not allowing anonymous users to help or view a task is set in allow_anonymous_users. Although, this is not yet applied.
> 
> Perhaps the keyword should change, because 'private' is indeed unclear. What do you think would be clearer? Something along the lines of PRIVATE_RESULT perhaps or the other way around, PUBLIC_RESULT.

PUBLIC_RESULT seems good. Then it can act like ALLOW_ANONYMOUS_USERS, true by default, but can be set to false. Also, I think a LISTED config should be added, so, if it is false, it cannot be found in the ViewTasks but only by getting the url.

**Comment by KaspervanM** [comment]
This PR is continued at #82.

**Pull request closed without merging**